active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,322 @@
+Libraries such as [yahoo/ember-intl](https://github.com/yahoo/ember-intl),
+[dockyard/ember-cli-i18n](https://github.com/dockyard/ember-cli-i18n), and
+[minutebase/ember-can](https://github.com/minutebase/ember-can) will be
+provided a viable public API to couple to.
+
+# Detailed design
+
+Stateful helpers have the same naming requirements, usage, and file placement
+as helpers today. They have similar lifecycle hooks to components.
+
+### Naming and file placement
+
+The requirement for naming a stateful helper are the same as those for existing
+helpers:
+
+* Helper names should use a `-`.
+* Helper names that do not use a `-` must be registered via
+  `Ember.Helpers.registerHelper`.
+* In Ember-CLI, helpers live in `app/helpers/full-name.js`, or with pods in
+  `app/full-name/helper.js`.
+
+### Defininition and lifecycle
+
+A stateful helper is defined as a class inheriting from `Ember.Helper`. For
+example:
+
+```js
+// app/helpers/hello-world.js
+import Ember from "ember";
+
+// Usage: {{hello-world}}
+export default Ember.Helper.extend({
+  compute() {
+    return "Hello Helper World";
+  }
+});
+```
+
+Upon initial render:
+
+* The helper instance is created.
+* The `compute` method is called. The return value is outputted where the
+  helper is used. For example in `<div class={{some-helper}}></div>` the return
+  value is set to the class. This is a slight change from current function
+  helper behavior, but the two behaviors should be aligned. See "Returning
+  a value" and "Block form" below.
+
+The `compute` function is always called with the same arguments a function
+helper is called with:
+
+```js
+// app/helpers/greet-someone.js
+import Ember from "ember";
+
+// Usage: {{greet-someone 'bob' greeting='say hello'}}
+export default Ember.Helper.extend({
+  compute(params, hash) {
+    return `Hello ${params[0]}, nice to ${hash.greeting}`;
+  }
+});
+```
+
+When the `params` or `hash` contents change, the `compute` method is called
+again. The instance of the helper is preserved across rerenders of the parent.
+
+The `init` and `destroy` methods can be subclassed for setup and teardown.
+
+`compute` is called with the same arguments as a function helper.
+
+### Consuming an stateful helper
+
+Stateful helpers can be used anywhere an HTMLBars subexpression can be
+used. For example:
+
+```hbs
+{{#if (can-access 'admin')}}
+  {{link-to 'login'}}
+{{/if}}
+{{my-login-button isAdmin=(can-access 'admin')}}
+<my-login-button isAdmin={{can-access 'admin'}} />
+Can access? {{can-access 'admin'}}
+```
+
+In all these cases, the helper is considered one-way. That is, it is a
+readable value and not two-way bound (toggling `isAdmin` does not impact the
+helper).
+
+Let's step through exactly what happens when using an helper like this:
+
+```hbs
+<my-login-button isAdmin={{can-access 'admin'}} />
+```
+
+Upon initial render:
+
+* The helper `can-access` is looked up on the container
```

Comment:
i suspect before this, it is looked up in the `helper` registry that `registerHelper` populates.